pilot/pkg/config/kube/ingress/controller.go

```diff
@@ -196,15 +196,13 @@ func (c *controller) onEvent(item types.NamespacedName) error {
 		Name:             item.Name + "-" + "virtualservice",
 		Namespace:        item.Namespace,
 		GroupVersionKind: gvk.VirtualService,
-		// Set this label so that we do not compare configs and just push.
-		Labels: map[string]string{constants.AlwaysPushLabel: "true"},
+		Annotations:      map[string]string{constants.InternalRouteSemantics: constants.RouteSemanticsIngress},
 	}
 	gatewaymetadata := config.Meta{
 		Name:             item.Name + "-" + "gateway",
 		Namespace:        item.Namespace,
 		GroupVersionKind: gvk.Gateway,
-		// Set this label so that we do not compare configs and just push.
```

Comment:
How is this fixed? I don't get it

Reply:
@howardjohn Take a look at the below logic, not only ingress but also gateway api go through this comparison
```
func needsPush(prev config.Config, curr config.Config) bool {

    ...
	// If the config is not Istio, let us just push.
	if !strings.HasSuffix(prev.GroupVersionKind.Group, "istio.io") {
		return true
	}
	// If current/previous metadata has "*istio.io" label/annotation, just push
	for label := range curr.Meta.Labels {
		if strings.Contains(label, "istio.io") {   
			return true
		}
	}
	for annotation := range curr.Meta.Annotations {
		if strings.Contains(annotation, "istio.io") {    // Here if we checked an annotation, we will always push
			return true
		}
	}
	for label := range prev.Meta.Labels {
		if strings.Contains(label, "istio.io") {
			return true
		}
	}
	for annotation := range prev.Meta.Annotations {
		if strings.Contains(annotation, "istio.io") {
			return true
		}
	}
	....
	return true
}

```

Reply:
wow that's extremely subtle and likely to break in the future. I really prefer the explicit AlwaysPush...

Reply:
Hmm, you know this stull of inner constant looks very hardcoding. I would like to remove all those by updating `configHandler` with an explicit arg, so we donot need to compare by `needPush` for some resources converted from ingress/gateway-api

Reply:
But currently we cannot gert rid of `InternalRouteSemantics` annotation easily, cause it is used across the code